examples/common.rs

```diff
@@ -0,0 +1,85 @@
+//! initialize and restore the terminal, and to install the panic and error hooks.
+//!
+//! The latest version of this code is available in the [examples] folder in the repository.
+//!
+//! Please note that the examples are designed to be run against the `main` branch of the Github
+//! repository. This means that you may not be able to compile with the latest release version on
+//! crates.io, or the one that you have installed locally.
+//!
+//! See the [examples readme] for more information on finding examples that match the version of the
+//! library you are using.
+//!
+//! [Ratatui]: https://github.com/ratatui-org/ratatui
+//! [examples]: https://github.com/ratatui-org/ratatui/blob/main/examples
+//! [examples readme]: https://github.com/ratatui-org/ratatui/blob/main/examples/README.md
+
+use std::{
+    io::{self, stdout, Stdout},
+    panic,
+};
+
+use color_eyre::{
+    config::{EyreHook, HookBuilder, PanicHook},
+    eyre::{self},
+};
+use crossterm::{
+    execute,
+    terminal::{
+        disable_raw_mode, enable_raw_mode, Clear, ClearType, EnterAlternateScreen,
+        LeaveAlternateScreen,
+    },
+};
+use ratatui::{backend::CrosstermBackend, Terminal};
+
+/// Initialize the terminal by enabling raw mode and entering the alternate screen.
+pub fn init_terminal() -> io::Result<Terminal<CrosstermBackend<Stdout>>> {
+    enable_raw_mode()?;
+    execute!(stdout(), EnterAlternateScreen)?;
+    let backend = CrosstermBackend::new(stdout());
+    let terminal = Terminal::new(backend)?;
+    Ok(terminal)
+}
+
+/// Restore the terminal by leaving the alternate screen and disabling raw mode.
+pub fn restore_terminal() -> color_eyre::Result<()> {
+    disable_raw_mode()?;
+    execute!(
+        stdout(),
+        LeaveAlternateScreen,
+        Clear(ClearType::FromCursorDown),
+    )?;
+    Ok(())
+}
+
+/// This replaces the standard `color_eyre` panic and error hooks with hooks that restore the
+/// terminal before printing the panic or error.
+pub fn install_hooks() -> color_eyre::Result<()> {
+    let (panic_hook, eyre_hook) = HookBuilder::default().into_hooks();
+    install_panic_hook(panic_hook);
+    install_error_hook(eyre_hook)?;
+    Ok(())
+}
+
+/// Install a panic hook that restores the terminal before printing the panic.
+fn install_panic_hook(panic_hook: PanicHook) {
+    // convert from a color_eyre PanicHook to a standard panic hook
+    let panic_hook = panic_hook.into_panic_hook();
+    panic::set_hook(Box::new(move |panic_info| {
+        let _ = restore_terminal();
+        panic_hook(panic_info);
+    }));
+}
+
+/// Install an error hook that restores the terminal before printing the error.
+fn install_error_hook(eyre_hook: EyreHook) -> color_eyre::Result<()> {
+    // convert from a color_eyre EyreHook to a standard eyre hook
+    let eyre_hook = eyre_hook.into_eyre_hook();
+    eyre::set_hook(Box::new(move |error| {
+        let _ = restore_terminal();
+        eyre_hook(error)
+    }))?;
+    Ok(())
+}
```

Comment:
We can, but I'm going to go with this being extremely useful based on seeing many different apps that crash for one reason or another (or return errors). When we don't use color_eyre or anyhow, then we have to specifically do:

```rust
let result = app.run();
restore_terminal()?;
result
```
Which is janky.

We're building a standard and opinionated approach here that many apps will build on top of. Ratatui is used by new rust users when they're learning rust, and it's useful to just do the right thing. Color eyre is a reasonable opinion that we can build on, is documented well in our website as a pattern to follow in the counter tutorial and the how-to pages.

This sets new users up to succeed automatically (when their app fails, they have a good way of understanding why upfront).

Reply:
This can work perfectly without an additional crate:

https://github.com/ratatui-org/ratatui/blob/9bd89c218afb1f3999dce1bfe6edea5b7442966d/examples/panic.rs#L52-L57

Reply:
As an aside, I think we probably should make some sort of easy way to integrate color_eyre into an app via crate feature flag instead of manually including it. E.g. perhaps something like:

```rust
ratatui::backend::crossterm::install_hooks()?;
let terminal = ratatui::backend::crossterm::terminal_on_stdout()?; // or something similar
```

(Note: the mechanism for enabling raw mode / entering the alternate screen is quite a bit different for each of the backends in ways that are generally incompatible with each other.)

I think you might be right that it would be nice to re-export crossterm instead of having to manually add that to your deps. Definitely a thing to consider outside of this PR though.

A concrete example of inconsistent approaches leading to user confusion can be found in the live streams that this newer rust user is running going through their learnings on Ratatui. https://www.youtube.com/@GoRustLang (see the recent livestreams - they're worth watching at 2x+ speed. In them the author actually hit questions about error handling and saw problems with it when they crash their app. They originally also learnt from the JSON tutorial, which takes a frame-centric approach to building apps rather than a widget approach. This made for confusion about how and where to get a buffer from when they were reading a widget-centric example. What this really showed me is that people learn non-linearly, and making sure there's consistency across the various examples and docs makes it easy for users to jump around and not be looking at drastic differences. It's something that I'd hypothesized that it would be a thing, but the series really showed that by example.

Reply:
I am in agreement about the non-linear learning and support consistency.

I'm personally a fan of having nicer panic hooks and have grown fond of the output from color-eyre. But also I see the value in showing only what is related to `ratatui`.

To that end, would having a `minimalist-example.rs` be useful?

Reply:
> This can work perfectly without an additional crate:

That's definitely far from "perfect":

- doesn't handle errors, only panics. A result return type is easy to overlook and forget to handle in the main method (e.g. just naively using the `?` operator)
- doesn't provide error messages that are anywhere close to those provided out of the box by color_eyre
- doesn't provide an abstraction that makes it easy to avoid talking about boxing errors
- error messages have the failing line of code somewhere in the stack trace, but require searching to work out the cause instead of being the first element

The benefits of color-eyre are all the sorts of things that help newer users fall into the pit of success rather than being tripped up.

Compare:
standard panic hook
<img width="1178" alt="image" src="https://github.com/ratatui-org/ratatui/assets/381361/7052fad1-ea27-4bad-8fc4-a25ec5c77786">

color-eyre panic hook
<img width="1196" alt="image" src="https://github.com/ratatui-org/ratatui/assets/381361/b94cec7b-41b1-4e06-90da-b59118880a36">

> One example for color_eyre, fine, all examples? Should be ratatui specific.

One common module used by all examples... this will be handled once, consistently the same way for all examples. There's no overhead to learning this or downside to it either.

```rust
mod common;

fn main() -> color_eyre::Result {
    common::install_hooks()?;
}
```

Reply:
> To that end, would having a `minimalist-example.rs` be useful?

Yup

Reply:
https://github.com/ratatui-org/ratatui/pull/1114

Reply:
My point here is that a panic hook can be achieved with std tools without additional dependencies.

Having a panic hook is definitely a good thing for examples. I just don’t see the benefit of color_eyre. Especially since the panic message formatting was improved recently.

Personal opinion: either something is bundled with ratatui, like reexported, then it’s fine to be included in common examples. Or it should stay in a single example on how to use ratatui with color_eyre instead of the std lib way. Ratatui depends on std, but it does not on color_eyre.

I dislike examples that mash different things together as a default so users don’t learn the parts. Instead it’s a bundled up bunch hard to separate. In this case it hides the fact that a panic hook is working perfectly fine without color_eyre.
The common example should be a minimal example what’s needed for a general app. Additional things should add on to it. Having colored panic messages is an additional thing. Exactly like the bar chart is not part of the common file.

Reply:
> My point here is that a panic hook can be achieved with std tools without additional dependencies.

> Having a panic hook is definitely a good thing for examples. I just don’t see the benefit of color_eyre. Especially since the panic message formatting was improved recently.

The message format pales in comparison to the color eyre format. Color eyre has benefits listed above. Those benefits are useful to writing apps.

> Personal opinion: either something is bundled with ratatui, like reexported, then it’s fine to be included in common examples. Or it should stay in a single example on how to use ratatui with color_eyre instead of the std lib way. Ratatui depends on std, but it does not on color_eyre.

Opinion: examples should use things which are widely used in the rust economy system. Color eyre and anyhow which it is based on have a much more significant reach compared to Ratatui.

> I dislike examples that mash different things together as a default so users don’t learn the parts. Instead it’s a bundled up bunch hard to separate. In this case it hides the fact that a panic hook is working perfectly fine without color_eyre.

You have to write a panic hook and learn how to do that, this is just the best available version of that.

> The common example should be a minimal example what’s needed for a general app. Additional things should add on to it. Having colored panic messages is an additional thing. Exactly like the bar chart is not part of the common file.

To summarize your perspetive on this:
- you believe that only necessary things to demonstrate the library calls should be in examples
- you believe color eyre provides no benefit over the built in panic hooks and messages that are printed on crashes
- you believe that adding color eyre makes examples harder to understand

I disagree with all three statements. If I've straw-manned your arguments, that's not my intent.

My rationale for this is based on looking at a wide variety of apps built on ratatui, many of them "writing an app to learn rust", and seeing what people end up with. Example code is copied verbatim **regularly**. I've also interacted and provide support with ratatui users across:
- discord (4 different servers)
- Reddit
- Rust lang forums
- Many code repos looking at common code patterns, bugs, pain points, issues etc.
- Issues / discussion in the ratatui repo

Each of these have shown the need for some common patterns to help drive apps to be better.

I've also written a significant amount of docs for Ratatui and read the feedback from users about those docs.

People learn non linearly and hit examples directly for the thing they care about. Having different approaches per example leads to the good approaches being skipped when they're only covered in a single place. Doing the right thing across all the examples is the right thing to do here.

Thinking more about this, I think it's probably worth bringing the common module into each example rather than extracting it to a separate file, as this makes each example self contained even at the cost of duplication.

I believe strongly that using color eyre in all our examples is the right approach to use here for ratatui, even taking into consideration and understanding your opinions to the contrary.

On the opposite note, I believe you were right about not using the prelude in example code and that's something I'd like to roll out completely as part of revamping the examples. Not sure if you saw the message I sent you about the forum post I made on the rust users forum about this https://users.rust-lang.org/t/to-use-prelude-or-to-not-to-use-prelude-that-is-the-question/110855. The opinions were divided there, but there was enough evidence that using the prelude hampered the understanding of the code for readers to make that change.